Function to write images to db & corrected writing comments to db.

joohing opened the pull request:
function to write images to the database in the image controller. Added a caption field to the post struct. Changed all varchar(255) types to text types since that apparently means nothing for performance. Implemented updating the list of comments tied to a post in the database by using text[] data type - so now the posts can have a variable amt of comments.

Commit:
Corrected errors in the queries in the comment controller. Wrote a
function to write images to the database in the image controller. Added
a caption field to the post struct. Changed all varchar(255) types to
text types since that apparently means nothing for performance.
Implemented updating the list of comments tied to a post in the database
by using text[] data type - so now the posts can have a variable amt of
comments.

Commit:
First attempt at querybuilder & reading from db.
First attempt at actually getting row data from the database and casting
to usable rust types.

joohing commented:
Absolut INGEN KRITIK TAK!

RummusBoi commented on backend/src/database/image_controller.rs:
```diff
@@ -1,35 +1,85 @@
 }
+
+async fn read_images_from_db(ids: Vec<String>) -> Result<Vec<Image>, sqlx::Error> {
+    let pool = PgPoolOptions::new().max_connections(5).connect(DATABASE_URL).await?;
```
Her kan vi bare bruge database_url fra env_vars

joohing replied:
Bitch

RummusBoi commented on backend/src/database/image_controller.rs:
```diff
@@ -1,35 +1,85 @@
+    let pool = PgPoolOptions::new().max_connections(5).connect(DATABASE_URL).await?;
+
+    let mut querybuilder_stub = QueryBuilder::new("select * from images where ");
+    let finalized_querybuilder = querybuilder_stub.push(ids.join(" or "));
```
Når vi joiner id'er sammen sådan her, har vi så stadig O(1) lookup time på hver række?

RummusBoi commented on backend/src/database/image_controller.rs:
```diff
@@ -1,35 +1,85 @@
+    Ok(query_result.iter().map(
+        // Issue: sqlx only supports up to 64 bit integers in the "get" method (else it complains
+        // that trait bound is not satisfied)
+        |row| Image {id: row.get::<String, _>("id"), data: row.get::<Vec<u8>, _>("data"), timestamp: row.get::<i64, _>("timestamp") as u128}
```
Hvad sker der hvis typerne i tabellen ikke passer her

RummusBoi replied:
der hvor vi typecaster i64 til u128 kan der i hvert fald opstå problemer. Dette program:
`
println!("{}", (-1 as i64) as u128);
`
Outputter 
`
340282366920938463463374607431768211455
`

RummusBoi replied:
Kan vi ikke bruge u64 i stedet for i64

joohing replied:
dsv ikke, vi bliver nok nødt til at validere værdierne på en anden måde (trait bound u64 not satisfied står der når man skriver u64)

RummusBoi commented on backend/src/database/image_controller.rs:
```diff
@@ -1,35 +1,85 @@
+
+// Better ways to do this?
+pub async fn generate_image_id(image_data: &Vec<u8>) -> Uuid {
+    return Uuid::new_v4();
```
Vi genererer efterhånden mange unikke id'er rundt omkring i koden, jeg synes vi skal lave en funktion i general_helpers der laver det for os, sådan at alle id'er genereres på samme måde

joohing replied:
True, måske også en måde som er hurtigere end at hashe? (Det er vist noget med at uuid bare er hash på Mac adresse og timestamp)

RummusBoi commented on backend/src/database/image_controller.rs:
```diff
@@ -1,35 +1,85 @@
+}
+
+// Better ways to do this?
+pub async fn generate_image_id(image_data: &Vec<u8>) -> Uuid {
```
image_data bliver ikke brugt her

RummusBoi replied:
Nvm vi bruger faktisk slet ikke denne funktion

joohing replied:
Tænkte man kunne bruge dataen til at generere et uuid med

RummusBoi commented on backend/src/database/image_controller.rs:
```diff
@@ -1,35 +1,85 @@
+    let pool = PgPoolOptions::new().max_connections(5).connect(DATABASE_URL).await?;
+
+    let mut querybuilder_stub = QueryBuilder::new("select * from images where ");
+    let finalized_querybuilder = querybuilder_stub.push(ids.join(" or "));
```
Er ikke lige inde i den her querybuilder ting, men vil disse to linjer ikke resultere i `"select * from images where id1 or id2 or id3"`

RummusBoi replied:
Og det er vel ikke en gyldig query

joohing replied:
det er en gyldig query, men der mangler et semikolon 🥲

RummusBoi replied:
men skal du ikke definere hvilket felt der er tale om?

Commit:
Improved error handling in write_image_to_db
Made a bunch of improvements including making row.get -> row.try_get,
and checking for underflow of the i64 when converting to u128.
